src/config.c

```diff
@@ -687,63 +632,151 @@ void loadServerConfig(char *filename, char config_from_stdin, char *options) {
+
+    /* Make sure we have an even number of arguments: conf-val pairs */
+    if (c->argc % 1) {
+        addReplyErrorFormat(c,"Unexpected CONFIG SET argument: %s", (char*)c->argv[c->argc-1]->ptr);
+        return;
+    }
+    config_count = (c->argc - 2) / 2;
 
-    /* Iterate the configs that are standard */
-    for (standardConfig *config = configs; config->name != NULL; config++) {
-        if (!(config->flags & IMMUTABLE_CONFIG) && 
-            (!strcasecmp(c->argv[2]->ptr,config->name) ||
-            (config->alias && !strcasecmp(c->argv[2]->ptr,config->alias))))
-        {
-            if (config->flags & SENSITIVE_CONFIG) {
-                redactClientCommandArgument(c,3);
-            }
+    set_configs = zcalloc(sizeof(standardConfig*)*config_count);
+    new_values = zmalloc(sizeof(sds*)*config_count);
+    old_values = zcalloc(sizeof(sds*)*config_count);
+    apply_fns = zcalloc(sizeof(apply_fn)*config_count);
 
-            if (config->flags & MULTI_ARG_CONFIG) {
-                argv = sdssplitlen(o->ptr, sdslen(o->ptr), " ", 1, &argc);
-                if (argv == NULL) {
-                    goto badfmt;
+    /* Find all relevant configs */
+    for (i = 0; i < config_count; i++) {
+        for (standardConfig *config = configs; config->name != NULL; config++) {
+            serverAssertWithInfo(c,c->argv[2+i*2],sdsEncodedObject(c->argv[2+i*2]));
+            serverAssertWithInfo(c,c->argv[2+i*2+1],sdsEncodedObject(c->argv[2+i*2+1]));
+            if (!(config->flags & IMMUTABLE_CONFIG) &&
+                (!strcasecmp(c->argv[2+i*2]->ptr,config->name) ||
+                 (config->alias && !strcasecmp(c->argv[2]->ptr,config->alias)))) {
+
+                if (config->flags & SENSITIVE_CONFIG) {
+                    redactClientCommandArgument(c,2+i*2+1);
                 }
-            } else {
-               argv = (char**)&o->ptr;
-               argc = 1;
-            }
 
-            if (!config->interface.set(config->data, argv, argc, 1, &errstr)) {
-                if (config->flags & MULTI_ARG_CONFIG) sdsfreesplitres(argv, argc);
-                goto badfmt;
+                /* If this config appears twice then fail */
+                for (j = 0; j < i; j++) {
+                    if (set_configs[j] == config) {
+                        /* Note: we don't abort the loop since we still want to handle redacting sensitive configs (above) */
+                        errstr = "duplicate parameter";
+                        invalid_args = 1;
+                        break;
+                    }
+                }
+                set_configs[i] = config;
+                new_values[i] = c->argv[2+i*2+1]->ptr;
+                break;
+            }
+        }
+        /* Fail if we couldn't find this config */
+        /* Note: we don't abort the loop since we still want to handle redacting sensitive configs (above) */
+        if (!set_configs[i]) {
+            errstr = "unrecognized parameter";
```

Comment:
No break here on purpose. See:
```c
* Note: we don't abort the loop since we still want to handle redacting sensitive configs (above) */
```

Reply:
It's tricky having a mutable error message since we're returning a static string pointer. It's possible doing this with a static (global) buffer. I saw we do this in at least once other similar case. But I don't like that idea that much. I think "unrecognized parameter" is pretty informative. I'll consider adding another message for the immutable config case.

Reply:
i suppose the easy way out is:
1. only report the right error message on the first error you encounter (but still have a specific error message for that issue).
2. keep running the loop just for the sake of redaction, but don't override the error message if already set.

in any case, i do think we wanna have specific error message, to let user figure out the problem faster.

Reply:
Fixed. Please re-review.

Reply:
Not sure if it's still relevant, but is having a static non-mutable string here so important? Saves some boilerplate code, but returning an allocated string might give us more flexibility.

Reply:
@yossigo I agree, but somehow I didn't change it during development and now I think it's not that relevant. Perhaps the next time we need some fancy error string we'll change this entire mechanism and add the extra flexibility.